ppafm/io.py

```diff
@@ -415,9 +416,56 @@ def loadGeometry(fname=None,format=None,params=None):
         if qs is None:
             qs = np.zeros(len(Zs))
         atoms = [list(Zs), list(xyzs[:, 0]), list(xyzs[:, 1]), list(xyzs[:, 2]), list(qs)]
-        nDim = params['gridN'].copy()
+        nDim = params["gridN"].copy()
 
-    return atoms,nDim,lvec
+    # Make sure lvec is a 4x3 array. Use default grid parameters if needed
+    if (lvec is None) or (len(lvec) == 0):
+        lvec = np.zeros((4, 3))
+        lvec[1, :] = params["gridA"].copy()
+        lvec[2, :] = params["gridB"].copy()
+        lvec[3, :] = params["gridC"].copy()
+    else:
+        if len(lvec) == 1:
+            lvec.append(params["gridA"].copy())
+        if len(lvec) == 2:
+            lvec.append(params["gridB"].copy())
+        if len(lvec) == 3:
+            lvec.append(params["gridC"].copy())
+    lvec = np.array(lvec)
+
+    # Shift scanning coordinates to such that actually make sense (being directly comparable to coordinates of atoms)
+    probeMin = params["scanMin"].copy()
+    probeMax = params["scanMax"].copy()
+    probeMin[0] += params["r0Probe"][0]
+    probeMin[1] += params["r0Probe"][1]
+    probeMin[2] -= params["r0Probe"][2]
+    probeMax[0] += params["r0Probe"][0]
+    probeMax[1] += params["r0Probe"][1]
+    probeMax[2] -= params["r0Probe"][2]
+
+    # Generate automatic lattice vectors and grid dimensions if needed
+    pad = 3.0
+    default_grid_step = 0.1
+    for i in range(3):
+        # Zero lattice vector is considered undefined and triggers creation of an automatic one.
+        # The automatic generated lattice vector should enclose the whole area filled with atoms as well as the whole scanning area, plus the default padding.
+        if np.allclose(lvec[i + 1, :], 0):
+            params["PBC"] = False
+            lvec[i + 1, i] = probeMax[i] + pad
```

Comment:
It seems to me like we are not taking into account the start of the scan region here. I think we would need to adjust the origin of the lvec to be at something like `probeMin - pad`, and then the length of the lattice vectors would be something like `probeMax - probeMin + 2 * pad`? Notice that currently `probeMin` is not used.

Reply:
Exactly right. The thing is, the current version of the code (at least the CPU part, I do not know about the OCL) does not work with non-zero origin in non-periodic systems. I should and I will solve this, but as a separate issue. In fact, I have added a line ``#lvec[i + 1, i] = probe_max[i] - probe_min[i] + 2*pad`` to `io.py`, which is commented-out for now but which I intend to use instead of the contemporary ``lvec[i + 1, i] = probe_max[i] + pad`` as soon as I solve the issue of an arbitrary origin.

Reply:
Ok, I didn't know that. It should work with any origin in the OCL code. I guess this by extension means that negative scan window coordinates don't work when `PBC = False`. Is there an issue for this? If not, then better make one so that it does not slip by.

Reply:
Exactly, last time I tried I got a blank area with zero forces for the part of scan region in which any of the three coordinates negative. There is no issue about it and I am going to make one. The original plan was to solve it as a by-product of #217 (may seem to be unrelated but as FHI-AIMS tends to be creative about the grid origin, it would have to be done to address that issue too). However, as I am being rather discouraged to implement #217, I will start a separate issue.